src/main/java/org/opensearch/neuralsearch/search/HybridTopScoreDocCollector.java

```diff
@@ -131,31 +129,34 @@ private TopDocs topDocsPerQuery(int start, int howMany, PriorityQueue<ScoreDoc>
 
         if (start < 0) {
             throw new IllegalArgumentException(
-                String.format(Locale.ROOT, "Expected value of starting position is between 0 and %d, got %d", size, start)
+                String.format(Locale.ROOT, "Expected value of starting position is between 0 and %d, got %d", howMany, start)
             );
         }
 
-        if (start >= size || howMany == 0) {
+        if (start >= howMany || howMany == 0) {
             return EMPTY_TOPDOCS;
         }
 
-        howMany = Math.min(size - start, howMany);
-        ScoreDoc[] results = new ScoreDoc[howMany];
+        int size = howMany - start;
+        ScoreDoc[] results = new ScoreDoc[size];
         // pq's pop() returns the 'least' element in the queue, therefore need
         // to discard the first ones, until we reach the requested range.
-        for (int i = pq.size() - start - howMany; i > 0; i--) {
+        for (int i = pq.size() - start - size; i > 0; i--) {
             pq.pop();
         }
 
         // Get the requested results from pq.
-        populateResults(results, howMany, pq);
+        populateResults(results, size, pq);
 
         return new TopDocs(new TotalHits(totalHits, totalHitsRelation), results);
     }
 
     protected void populateResults(ScoreDoc[] results, int howMany, PriorityQueue<ScoreDoc> pq) {
         for (int i = howMany - 1; i >= 0; i--) {
```

Comment:
```suggestion
        for (int i = howMany - 1; i >= 0 && i < results.length && pq.isEmpty() == false; i--) {
```

Reply:
I think it's not exactly same as we do have now, we may not even start the loop. E.g. if heap has elements and ```howMany = 10``` and ```results.length = 4```, then with existing logic we'll do 4 iterations, but with suggested change there will be none.
I'll add the heap size check though, it makes sense